core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+                    loop {
+                        if simulation_duration_since_base > required_duration_since_base {
+                            break;
+                        }
+                        let current_simulation_time = SystemTime::now();
+                        simulation_duration_since_base = current_simulation_time
+                            .duration_since(base_simulation_time)
+                            .unwrap();
+                    }
+
+                    let sender = match label {
+                        ChannelLabel::NonVote => &non_vote_sender,
+                        ChannelLabel::TpuVote => &tpu_vote_sender,
+                        ChannelLabel::GossipVote => &gossip_vote_sender,
+                        ChannelLabel::Dummy => unreachable!(),
+                    };
+                    sender.send(batches_with_stats.clone()).unwrap();
+
+                    let batches = &batches_with_stats.0;
+                    let (batch_count, tx_count) = (
+                        batches.len(),
+                        batches.iter().map(|batch| batch.len()).sum::<usize>(),
+                    );
+                    debug!(
+                        "sent {:?} {} batches ({} txes)",
+                        label, batch_count, tx_count
+                    );
+                    let (total_batch_count, total_tx_count) = match label {
+                        ChannelLabel::NonVote => (&mut non_vote_count, &mut non_vote_tx_count),
+                        ChannelLabel::TpuVote => (&mut tpu_vote_count, &mut tpu_vote_tx_count),
+                        ChannelLabel::GossipVote => (&mut gossip_vote_count, &mut gossip_vote_tx_count),
+                        ChannelLabel::Dummy => unreachable!(),
+                    };
+                    *total_batch_count += batch_count;
+                    *total_tx_count += tx_count;
+
+                    let log_interval = simulation_duration_since_base - last_log_duration;
+                    if log_interval > Duration::from_millis(100) {
+                        let current_tx_count = non_vote_tx_count + tpu_vote_tx_count + gossip_vote_tx_count;
+                        let duration =log_interval.as_secs_f64();
+                        let tps = (current_tx_count - last_tx_count) as f64 / duration;
+                        let non_vote_tps = (non_vote_tx_count - last_non_vote_count) as f64 / duration;
+                        let tpu_vote_tps = (tpu_vote_tx_count - last_tpu_vote_tx_count) as f64 / duration;
+                        let gossip_vote_tps = (gossip_vote_tx_count - last_gossip_vote_tx_count) as f64 / duration;
+                        info!(
+                            "senders(non-,tpu-,gossip-vote): tps: {} (={}+{}+{}) over {:?} not-recved: ({}+{}+{})",
+                            tps as u64,
+                            non_vote_tps as u64,
+                            tpu_vote_tps as u64,
+                            gossip_vote_tps as u64,
```

Comment:
just cast these above?

Reply:
well, i think this is more clear for its intention..: 619e7e58073dad0b1824cd296c4ac04716c61ea1